File: lib/khiva/matrix.rb
module Khiva
  module Matrix
    class << self
      def find_best_n_discords(profile, index, m, n, self_join: false)
        discord_distances = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        discord_indices = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        subsequence_indices = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:find_best_n_discords, profile, index, m, n, discord_distances, discord_indices, subsequence_indices, self_join ? 1 : 0)
        [Array.new(discord_distances), Array.new(discord_indices), Array.new(subsequence_indices)]
      end

      def find_best_n_motifs(profile, index, m, n, self_join: false)
        motif_distances = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        motif_indices = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        subsequence_indices = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:find_best_n_motifs, profile, index, m, n, motif_distances, motif_indices, subsequence_indices, self_join ? 1 : 0)
        [Array.new(motif_distances), Array.new(motif_indices), Array.new(subsequence_indices)]
      end

      def mass(q, t)
        distances = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:mass, q, t, distances)
        Array.new(distances)
      end

      def stomp(tssa, tssb, m)
        profile = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        index = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:stomp, tssa, tssb, m, profile, index)
        [Array.new(profile), Array.new(index)]
      end

      def stomp_self_join(tss, m)
        profile = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        index = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:stomp_self_join, tss, m, profile, index)
        [Array.new(profile), Array.new(index)]
      end

      def matrix_profile(tssa, tssb, m)
        profile = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        index = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:matrix_profile, tssa, tssb, m, profile, index)
        [Array.new(profile), Array.new(index)]
      end

      def matrix_profile_self_join(tss, m)
        profile = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        index = Fiddle::Pointer.malloc(Fiddle::SIZEOF_VOIDP)
        FFI.call(:matrix_profile_self_join, tss, m, profile, index)
        [Array.new(profile), Array.new(index)]
      end
    end
  end
end

File: test/support/matrix.py
from khiva.array import *
from khiva.library import *
from khiva.matrix import *

set_backend(KHIVABackend.KHIVA_BACKEND_CPU)

a1 = Array.from_list([11, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 11], dtype.s32)
a2 = Array.from_list([9, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 9], dtype.s32)
stomp_result = stomp(a1, a2, 3)

find_best_n_discords_result = find_best_n_discords(stomp_result[0], stomp_result[1], 3, 2)

print('find_best_n_discords_result')
print(find_best_n_discords_result[0].to_list())
print(find_best_n_discords_result[1].to_list())
print(find_best_n_discords_result[2].to_list())
print()

find_best_n_motifs_result = find_best_n_motifs(stomp_result[0], stomp_result[1], 3, 2)

print('find_best_n_motifs_result')
print(find_best_n_motifs_result[0].to_list())
print(find_best_n_motifs_result[1].to_list())
print(find_best_n_motifs_result[2].to_list())
print()

print('mass')
print(mass(a1, a2).to_list())
print()

print('stomp')
res = stomp(a1, a2, 3)
print(res[0].to_list())
print(res[1].to_list())
print()

print('stomp_self_join')
res = stomp_self_join(a1, 3)
print(res[0].to_list())
print(res[1].to_list())
print()

print('matrix_profile')
res = matrix_profile(a1, a2, 3)
print(res[0].to_list())
print(res[1].to_list())
print()

print('matrix_profile_self_join')
res = matrix_profile_self_join(a1, 3)
print(res[0].to_list())
print(res[1].to_list())
print()
